port_ocean/helpers/metric.py

```diff
@@ -0,0 +1,173 @@
+    OBJECT_COUNT = "object_count"
+    DURATION = "duration"
+    ERROR_COUNT = "error_count"
+    INPUT_COUNT = "input_count"
+
+
+class MetricType:
+    EXTRACT = "extract"
+    TRANSFORM = "transform"
+    LOAD = "load"
+
+
+class MetricAggregator:
+    def __init__(self) -> None:
+        self._lock = asyncio.Lock()
+        self.metrics: KindsMetricsData = KindsMetricsData()
+
+    async def get_metrics(self) -> dict[str, MetricsData]:
+        return self.metrics.metrics
+
+    async def get_metric(self) -> TransformStats | LoadStats | ExtractStats | None:
+        phase = port_ocean.context.event.event.attributes.get("phase", None)
+        if not phase:
+            return None
+
+        metric = self.metrics.metrics.get(resource.resource.kind)
+        if not metric:
+            self.metrics.metrics[resource.resource.kind] = MetricsData()
+
+        return getattr(self.metrics.metrics.get(resource.resource.kind), phase)
+
+    async def increment_field(self, field: str, amount: int | float = 1) -> None:
+        metric = await self.get_metric()
+        async with self._lock:
+            val = getattr(metric, field)
+            metric.__setattr__(field, val + amount)
+
+    async def increment_status(self, status_code: str) -> None:
+        metric = await self.get_metric()
+        if metric is None or not isinstance(metric, ApiStats):
+            return None
+        async with self._lock:
+
+            status = metric.requests.get(status_code)
+            if not status:
+                metric.requests[status_code] = 0
+            metric.requests[status_code] = metric.requests.get(status_code, 0) + 1
+
+    async def flush(self) -> None:
+        async with self._lock:
+            metric_dict = asdict(self.metrics)
+            logger.info(f"integration metrics {json.dumps(metric_dict)}")
+        await self.reset()
+
+    async def reset(self) -> None:
+        async with self._lock:
+            self.metrics = KindsMetricsData()
+
+
+async def timed_generator(
+    generator: ASYNC_GENERATOR_RESYNC_TYPE,
+) -> ASYNC_GENERATOR_RESYNC_TYPE:
+    if not port_ocean.context.event.event.should_record_metrics:
+        async for items in generator:
+            yield items
+    async with port_ocean.context.event.event_context(
+        port_ocean.context.event.EventType.METRIC, attributes={"phase": "extract"}
+    ):
+        while True:
+            try:
+                start = time.monotonic()
+                items = await anext(generator)
+                end = time.monotonic()
+                duration = end - start
+                await port_ocean.context.event.event.increment_metric(
+                    MetricFieldType.DURATION, duration
+                )
+                await port_ocean.context.event.event.increment_metric(
+                    MetricFieldType.OBJECT_COUNT, len(items)
+                )
+
+                yield items
+            except Exception:
+                break
```

Comment:
why `break`? how the `exception` is being propogated? its not equivalent to the way you are iterating over  when `if not port_ocean.context.event.event.should_record_metrics`